cli/report/report.go

```diff
@@ -0,0 +1,184 @@
+	}
+	outJSON, _ := json.MarshalIndent(wrapped, "", "  ")
+	rawAssetFileName = "raw_assets_" + fileMidName + ".json"
+	err = ioutil.WriteFile(filepath.Join(reportOutputPath, rawAssetFileName), outJSON, 0644)
+	if err != nil {
+		return "", err
+	}
+	return
+}
+
+func generateReportData(rawAssetFileName string, reportQueryPath string, reportOutputPath string) (results interface{}, err error) {
+	// Load resources from json and rego files
+	resources, err := loader.All([]string{filepath.Join(reportOutputPath, rawAssetFileName), reportQueryPath})
+	if err != nil {
+		return nil, err
+	}
+	compiler, err := resources.Compiler()
+	if err != nil {
+		return nil, err
+	}
+	store, err := resources.Store()
+	if err != nil {
+		return nil, err
+	}
+	r := rego.New(
+		rego.Query(`data.reports`),
+		rego.Compiler(compiler),
+		rego.Store(store),
+	)
+	rs, err := r.Eval(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	results = rs[0].Expressions[0].Value
+
+	return results, nil
+}
+
+func printReports(results interface{}, reportOutputPath string, format string, fileSuffix string) error {
+	resultsMap := results.(map[string]interface{})
+	for group, reports := range resultsMap {
+		reportsMap := reports.(map[string]interface{})
+		for reportName, content := range reportsMap {
+			if strings.HasSuffix(reportName, "_report") {
+				reportFileName := group + "." + reportName + "_" + fileSuffix
+				fmt.Println("Printing " + group + "." + reportName)
+				if format == "json" {
+					reportFileName = reportFileName + ".json"
+					fileContent, err := json.MarshalIndent(content, "", "  ")
+					if err != nil {
+						return err
+					}
+					err = ioutil.WriteFile(filepath.Join(reportOutputPath, reportFileName), fileContent, 0644)
+					if err != nil {
+						return err
+					}
+				} else {
+					reportFileName = reportFileName + ".csv"
+					contentSlice := content.([]interface{})
+					f, _ := os.Create(filepath.Join(reportOutputPath, reportFileName))
+
+					defer f.Close()
+					w := csv.NewWriter(f)
+					if len(contentSlice) > 0 {
+						firstRow := contentSlice[0]
+						var keys []string
+						firstRowMap := firstRow.(map[string]interface{})
+						for key := range firstRowMap {
+							keys = append(keys, key)
+						}
+						sort.Strings(keys)
+						w.Write(keys)
+						w.Flush()
+						for _, record := range contentSlice {
+							recordMap := record.(map[string]interface{})
+							var record []string
+							for _, key := range keys {
+								record = append(record, recordMap[key].(string))
+							}
+							w.Write(record)
+						}
+						w.Flush()
+					} else {
+						f, err := os.Create(filepath.Join(reportOutputPath, reportFileName))
+						if err != nil {
+							return err
+						}
+						defer f.Close()
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func ListAvailableReports(reportQueryPath string) error {
+	// Load resources from json and rego files
+	resources, err := loader.All([]string{reportQueryPath})
```

Comment:
This logic looks duplicative, I'd make a small suggestion to refactor so this and `generateReportData` share common underlying methods. In particular, logic like finding reports and knowing their suffix shouldn't be duplicated.

Non-blocking.

Reply:
done in https://github.com/GoogleCloudPlatform/cloud-foundation-toolkit/pull/315/commits/df61d7d7fd5592a35b28cfc9d0f10afb533a3aaf